# DataFrames Advanced Methods
![](title_pict/df_advanced2.png)

In [1]:
import pandas as pd

The following dataframes will be used throughout this section.

In [2]:
df1 = pd.DataFrame({'1C1':[3,4,5], '1C2':[30,40,50], 'col':[300,400,500]}, index=['A', 'B', 'C'])
df1

1C1  1C2  col
A    3   30  300
B    4   40  400
C    5   50  500

In [3]:
df2 = pd.DataFrame({'2C1':[5,6,9], '2C2':[50,60,90]}, index=['A', 'B', 'C'])
df2

2C1  2C2
A    5   50
B    6   60
C    9   90

In [4]:
df3 = pd.DataFrame({'3C1':[7,8], '3C2':[70,80], 'col':[700,800]}, index=['C', 'D'])
df3

3C1  3C2  col
C    7   70  700
D    8   80  800

## concat()
It is used to concatenate dataframes along an axis (horizontally or vertically).

In [5]:
# default axis=0: add as new rows
# two new rows coming from df3: C and D
pd.concat([df1,df3])        

1C1   1C2  col  3C1   3C2
A  3.0  30.0  300  NaN   NaN
B  4.0  40.0  400  NaN   NaN
C  5.0  50.0  500  NaN   NaN
C  NaN   NaN  700  7.0  70.0
D  NaN   NaN  800  8.0  80.0

In [6]:
# axis=1, add as new columns
# three more rows are coming from df3: 3C1, 3C2, col 
pd.concat([df1,df3], axis=1)   

1C1   1C2    col  3C1   3C2    col
A  3.0  30.0  300.0  NaN   NaN    NaN
B  4.0  40.0  400.0  NaN   NaN    NaN
C  5.0  50.0  500.0  7.0  70.0  700.0
D  NaN   NaN    NaN  8.0  80.0  800.0

In [7]:
# along only on common columns
pd.concat([df1,df2], join='inner')   # intersection of columns

Empty DataFrame
Columns: []
Index: [A, B, C, A, B, C]

- If the indexes of two dataframes are the same, the column labels are different, and axis=1, then the second dataframe is concatenated horizontally.

In [8]:
# default axis=1: add as a new row
pd.concat([df1,df2], axis=1) 

1C1  1C2  col  2C1  2C2
A    3   30  300    5   50
B    4   40  400    6   60
C    5   50  500    9   90

## shift()
It is used to shift the data up or down.

In [9]:
df = pd.DataFrame(['A', 'B', 'C', 'D', 'E'])
df

0
0  A
1  B
2  C
3  D
4  E

In [10]:
# shift the values down by 1 row
df.shift()

0
0  None
1     A
2     B
3     C
4     D

In [11]:
# shift the values down by 2 rows
df.shift(2)

0
0  None
1  None
2     A
3     B
4     C

In [12]:
# shift the values up by 1 row
df.shift(-1)

0
0     B
1     C
2     D
3     E
4  None

In [13]:
# shift the values up by 2 rows
df.shift(-2)

0
0     C
1     D
2     E
3  None
4  None

## pct_change()
Percentage change between the current and a prior value.

In [14]:
df = pd.DataFrame([500, 400, 600, 150, 180])
df

0
0  500
1  400
2  600
3  150
4  180

In [15]:
df.pct_change()

0
0   NaN
1 -0.20
2  0.50
3 -0.75
4  0.20

## rolling()
It is used to do calculations using a rolling window.
- The most commonly used methods following the *rolling()* method are *sum()*, *mean()*, *median()*, and *std()*.

In [16]:
df

0
0  500
1  400
2  600
3  150
4  180

- rolling() returns a Rolling object.
- The *window* parameter specifies the size of the moving window and must be provided.

In [23]:
# window=3
type(df.rolling(3))

pandas.core.window.rolling.Rolling

In [24]:
df.rolling(window=3).sum()

0
0     NaN
1     NaN
2  1500.0
3  1150.0
4   930.0

With a moving window size of 3, the row triplets and corresponding values are as follows:
- Rows 0, 1, and 2: The sum of 500, 400, and 600 is 1500.
- Rows 1, 2, and 3: The sum of 400, 600, and 150 is 1150.
- Rows 2, 3, and 4: The sum of 600, 150, and 180 is 930.
  
The first two rolling sum values are NaN (not a number) because a minimum of 3 values is required to compute the sum